cloud/scope/powervs_cluster_test.go

```diff
@@ -1704,3 +1704,515 @@ func TestCreateServiceInstance(t *testing.T) {
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{Spec: infrav1beta2.IBMPowerVSClusterSpec{
+				VPC:           &infrav1beta2.VPCResourceReference{ID: ptr.To("VPCID")},
+				ResourceGroup: &infrav1beta2.IBMPowerVSResourceReference{ID: ptr.To("resourceGroupID")}}},
+		}
+		vpcOutput := &vpcv1.VPC{Name: ptr.To("VPCName"), ID: ptr.To("VPCID")}
+		mockVPC.EXPECT().GetVPC(gomock.Any()).Return(vpcOutput, nil, nil)
+
+		output, err := clusterScope.ReconcileVPC()
+		g.Expect(err).To(BeNil())
+		g.Expect(output).To(BeFalse())
+	})
+
+	t.Run("When VPC ID is already set and GetVPC returns error", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{Spec: infrav1beta2.IBMPowerVSClusterSpec{
+				VPC:           &infrav1beta2.VPCResourceReference{ID: ptr.To("VPCID")},
+				ResourceGroup: &infrav1beta2.IBMPowerVSResourceReference{ID: ptr.To("resourceGroupID")}}},
+		}
+		mockVPC.EXPECT().GetVPC(gomock.Any()).Return(nil, nil, fmt.Errorf("GetVPC returns error"))
+		output, err := clusterScope.ReconcileVPC()
+		g.Expect(err).ToNot(BeNil())
+		g.Expect(output).To(BeFalse())
+	})
+
+	t.Run("When VPC ID is already set and GetVPC returns empty output", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Spec: infrav1beta2.IBMPowerVSClusterSpec{
+					//VPC:           &infrav1beta2.VPCResourceReference{ID: ptr.To("VPCID")},
+					ResourceGroup: &infrav1beta2.IBMPowerVSResourceReference{ID: ptr.To("resourceGroupID")}},
+				Status: infrav1beta2.IBMPowerVSClusterStatus{VPC: &infrav1beta2.ResourceReference{ID: ptr.To("VPCID")}},
+			},
+		}
+		mockVPC.EXPECT().GetVPC(gomock.Any()).Return(nil, nil, nil)
+		output, err := clusterScope.ReconcileVPC()
+		g.Expect(err).ToNot(BeNil())
+		g.Expect(output).To(BeFalse())
+	})
+
+	t.Run("When VPC ID is already set and GetVPC returns vpc in pending state", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+
+		vpcOutput := &vpcv1.VPC{Name: ptr.To("VPCName"), ID: ptr.To("VPCID"), Status: ptr.To(string(infrav1beta2.VPCStatePending))}
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{Spec: infrav1beta2.IBMPowerVSClusterSpec{
+				//VPC:           &infrav1beta2.VPCResourceReference{ID: ptr.To("VPCID")},
+				ResourceGroup: &infrav1beta2.IBMPowerVSResourceReference{ID: ptr.To("resourceGroupID")}},
+				Status: infrav1beta2.IBMPowerVSClusterStatus{VPC: &infrav1beta2.ResourceReference{ID: ptr.To("VPCID")}},
+			},
+		}
+		mockVPC.EXPECT().GetVPC(gomock.Any()).Return(vpcOutput, nil, nil)
+		output, err := clusterScope.ReconcileVPC()
+		g.Expect(err).To(BeNil())
+		g.Expect(output).To(BeTrue())
```

Comment:
Is validating vpc id set in status is also checked?

Reply:
There is no status set [here](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/cloud/scope/powervs_cluster.go#L1004), should i check for nil instead?

Reply:
oh ok, not sure why we are not setting status, In other resources we are setting status right, when spe.id is set?

Reply:
so since you are only setting the status, You can ignore that check.